garak/configurable.py

```diff
@@ -0,0 +1,82 @@
+        # plugins['detectors'][x]['generators']['rest']['RestGenerator']
+        # plugins['detectors'][x]['generators']['rest']
+        # plugins['probes'][y]['generators']['rest']['RestGenerator']
+        if len(namespace_parts) > 2:
+            # example class expected garak.generators.huggingface.Pipeline
+            # spec_type = generators
+            # namespace = huggingface
+            # classname = Pipeline
+            # current expected spec_type values are `_plugins.PLUGIN_TYPES`
+            spec_type = namespace_parts[-2]
+            namespace = namespace_parts[-1]
+            classname = self.__class__.__name__
+            plugins_config = {}
+            if isinstance(local_root, dict) and spec_type in local_root:
+                plugins_config = local_root[spec_type]
+            elif hasattr(local_root, spec_type):
+                plugins_config = getattr(local_root, spec_type)
+            if namespace in plugins_config:
+                # example values:
+                # generators: `nim/openai/huggingface`
+                # probes: `dan/gcg/xss/tap/promptinject`
+                attributes = plugins_config[namespace]
+                namespaced_klass = f"{namespace}.{classname}"
+                self._apply_config(attributes)
+                if classname in attributes:
+                    self._apply_config(attributes[classname])
+                elif namespaced_klass in plugins_config:
+                    # for compatibility remove after
+                    logging.warning(
+                        f"Deprecated configuration key found: {namespaced_klass}"
+                    )
+                    self._apply_config(plugins_config[namespaced_klass])
+        self._apply_missing_instance_defaults()
+        self.loaded = True
+
+    def _apply_config(self, config):
+        classname = self.__class__.__name__
+        init_params = inspect.signature(self.__init__).parameters
+        for k, v in config.items():
+            if k in _plugins.PLUGIN_TYPES or k == classname:
+                # skip entries for more qualified items or any plugin type
+                # should this be coupled to `_plugins`?
+                continue
+            if hasattr(self, "_supported_params") and k not in self._supported_params:
+                # if the class has a set of supported params skip unknown params
+                # should this pass signature arguments as supported?
+                logging.warning(f"Unknown configuration key for {classname}: {k}")
+                continue
+            if hasattr(self, k):
+                # do not override values provide by caller that are not defaults
+                if k in init_params and (
+                    init_params[k].default is inspect.Parameter.empty
+                    or (
+                        init_params[k].default is not inspect.Parameter.empty
+                        and getattr(self, k) != init_params[k].default
```

Comment:
This is the `nuance` referred to in the PR description, I am open to ideas on how to this might be adjusted, there might be some way better way to mark the value already in the object as having been passed explicitly vs defaulted by from the constructor. For now that edge case might deserve a clear test case that show the current behavior.

Reply:
This is actually why the the config should be passed in, the global a `default` value if the user provides a `config_root` that is not the shared global object the instantiation will use that provided config.